### DISCUSSION TITLE
CAN<->Ethernet ANDL error

### DISCUSSION THREAD

**Discussion by pminjiq** [CoRE-RG/SignalsAndGateways#6] [Q&A] [answered]
Hello
I have a problem to make the ANDL when I set CAN<->Ethernet as below.
![image](https://user-images.githubusercontent.com/105440387/179177057-8dcbe475-6640-44a6-b6ca-aed0ff355d53.png)
![image](https://user-images.githubusercontent.com/105440387/179177430-4f02ae3a-1ed4-48ab-875a-7f2874ba72d6.png)

I saw this SignalAndGateways can support CAN<->Ethernet communication. 
So, my question is what should I wrote in the ANDL to make CAN<->Ethernet communication possible. 

Thanks for your help

Minji

**Comment by SaloM0nk** [comment]
Hello Minji, 

I am no expert with our ANDL, sadly. I assigned Philipp to this problem, but he is currently on not available and will be back in a few weeks. 

For now: 
Please check out the examples in SignalsAndGateawys -> examples_andl.
It seems that you try to receive the message on an Ethernet node which might not be supported by the ANDL. I think (and might well be wrong) that the ANDL can only calculate the path from  CANNode -> Gateway -> ETHERNET -> Gateway -> CANNode. 
SignalsAndGateways for sure supports CANNode -> Gateway -> ETHERNET -> Ethernet Node, but you might have to configure this by hand. In the Gateway XML file you can set the Ethernet transformation for each CAN ID to be forwarded by the gateway to a specific Ethernet destination MAC. 

Hope this helps. If not please provide more information about your andl network definition and maybe Philipp can help when he is back. 

Regards,
Timo

**Comment by pminjiq** [comment]
Hi Timo 

Thanks for your reply

I checked the example in SignalsAndGateawys. but, there is only CAN<->CAN example.
So I tried to configure Gateway XML file to communicate Ethernet Node -> Gateway -> ETHERNET -> CAN Node. but it didn't work. (It didn't transform to CAN)
![image](https://user-images.githubusercontent.com/105440387/179446588-d4311d81-a26e-4b3c-b939-242a315c5051.png)
![image](https://user-images.githubusercontent.com/105440387/179446611-8d620e48-1f48-47e1-bd5e-d28919e582b8.png)

I configured the Etherent frame is Best effort and  the Gateway to connect with can bus and destination can node as below.
![image](https://user-images.githubusercontent.com/105440387/179337270-76a7e854-7361-4c50-8833-d8a63cb3d5c8.png)

Please tell me if I wrote something wrong or need to write more...
Also, I need to know how to communicate CANNode -> Gateway -> ETHERNET -> Ethernet Node. Could you show me a example..?

Thanks a lot!

Minji

**Comment by SaloM0nk** [comment, marked as answer]
Hi Minji,

For a CANNode -> Gateway -> ETHERNET -> Ethernet Node I have made the following modification on the small network example using the `WithQ `config:
1. disabled buffering in the omnetpp.ini line 45 `**.gateway.buffering[*].typename = "NoGatewayBuffering"`
2. removed canNode2, canbus2, and gateway2 from smallNetwork.ned with all their connections.
3. added EthernetSink Node
    * add import for host `import core4inet.nodes.ethernet.tsn.TSNEtherHost;`
    * add node to submodules `ethSink: TSNEtherHost;`
    * add connections: `ethSink.ethg <--> cable {  delay = 50ns; } <--> switch.ethg[1];`
4. added ethSink ini: 
     `**.ethSink.phy[*].mac.address = "0A-00-00-00-00-04"`
     `**.ethSink.phy[*].untaggedVID = 0`
     `**.ethSink.phy[*].taggedVIDs = "0,1"`
     `**.ethSink.numApps = 1`
     `**.ethSink.app[0].typename = "IEEE8021QTrafficSinkApp"`
     `**.ethSink.app[0].srcAddress = "0A-00-00-00-00-03"`
     `**.ethSink.app[0].replyFirst = false`
     `**.ethSink.bgIn.destination_gates = "app[0].in"`

When I run the network for 1 second all 217 transformed CAN messages (canId 4 is not transformed in the example) arrive at the application. 

I don't know your network but the gateway configuration looks fine. I think the problem is the same as I describe below.

For the other direction it is not that simple since we do not have any Ethernet source apps that send encapsulated CAN frames. You could write your own source app, e.g., extending the `IEEE8021QTrafficSourceApp` from CoRE4INET, so that it sends exactly the same encapsulated CAN frames as created by the `GatewayTransformation::transformCanFrame`. 

Regards,
Timo

**Comment by pminjiq** [comment]
Hi Timo

Thank you so much. it worked! 
I really appreciate your help.

Regards,
Minji
